x-pack/platform/test/spaces_api_integration/common/plugins/read_only_objects_test_plugin/server/plugin.ts

```diff
@@ -0,0 +1,375 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type { CoreSetup, Plugin } from '@kbn/core/server';
+
+const READ_ONLY_TYPE = 'read_only_type';
+const NON_READ_ONLY_TYPE = 'non_read_only_type';
+
+export class ReadOnlyObjectsPlugin implements Plugin {
+  public setup(core: CoreSetup) {
+    core.savedObjects.registerType({
+      name: READ_ONLY_TYPE,
+      hidden: false,
+      namespaceType: 'multiple-isolated',
+      supportsAccessControl: true,
+      mappings: {
+        dynamic: false,
+        properties: {
+          description: { type: 'text' },
+        },
+      },
+    });
+
+    core.savedObjects.registerType({
+      name: NON_READ_ONLY_TYPE,
+      hidden: false,
+      namespaceType: 'multiple-isolated',
+      mappings: {
+        dynamic: false,
+        properties: {
+          description: { type: 'text' },
+        },
+      },
+    });
+
+    const router = core.http.createRouter();
+
+    router.post(
+      {
+        path: '/read_only_objects/create',
+        security: {
+          authz: {
+            enabled: false,
+            reason: 'This route is opted out from authorization',
+          },
+        },
+        validate: {
+          body: schema.object({
+            type: schema.maybe(
+              schema.oneOf([schema.literal(READ_ONLY_TYPE), schema.literal(NON_READ_ONLY_TYPE)])
+            ),
+            isReadOnly: schema.maybe(schema.boolean()),
+          }),
+        },
+      },
+      async (context, request, response) => {
+        const soClient = (await context.core).savedObjects.getClient();
+        const objType = request.body.type || READ_ONLY_TYPE;
+        const { isReadOnly } = request.body;
+
+        const options = isReadOnly ? { accessControl: { accessMode: 'read_only' as const } } : {};
+        try {
+          const result = await soClient.create(
+            objType,
+            {
+              description: 'test',
+            },
+            options
+          );
+
+          return response.ok({
+            body: result,
+          });
+        } catch (error) {
+          return response.customError({
```

Comment:
Here and in the other routes...

By defining custom error responses we're always returning a 400 despite whatever error code the saved objects client returns.

The saved objects client returns hapi errors so that if you throw a conflict error from e.g. update, it will be caught be core's router and returned as a 409 conflict error.

By pinning the status code the tests don't actually test if e.g. an authorization failure returns a 403, it just tests your test plugin's code.